### ISSUE TITLE
Stop throwing first chance exceptions when CreateInstance is called on generic extension point types

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#523]
Very much like #494, only with open generic types.
When we try to instantiate argument formatters, fake configurators, and dummy factories, we may try to instantiate an open generic type, causing a first chance exception.

For an example, see [Configure_should_be_null_guarded](https://github.com/FakeItEasy/FakeItEasy/blob/dbfc0289f2493ee10d0aaf325fdc8734307645b4/Source/FakeItEasy.Tests/Core/FakeConfiguratorTests.cs#L25). We try to instantiate `TestableConfigurator<T>`.

**Comment by blairconrad** [comment]
I should've said, but the exception details (from "Copy exception detail to clipboard" ![image](https://cloud.githubusercontent.com/assets/3275797/11780436/e90eee88-a22f-11e5-8359-5ff10db73c30.png)) look like 


```
System.ArgumentException occurred
  HResult=-2147024809
  Message=Cannot create an instance of FakeItEasy.Tests.Core.FakeConfiguratorTests+TestableConfigurator`1[T] because Type.ContainsGenericParameters is true.
  Source=mscorlib
  StackTrace:
       at System.RuntimeType.CreateInstanceCheckThis()
       at System.RuntimeType.CreateInstanceSlow(Boolean publicOnly, Boolean skipCheckThis, Boolean fillCache, StackCrawlMark& stackMark)
       at System.Activator.CreateInstance(Type type, Boolean nonPublic)
       at System.Activator.CreateInstance(Type type)
       at FakeItEasy.Core.TypeCatalogueInstanceProvider.InstantiateAllOfType[T]() in c:\PortableApps\Documents\Source\GitHub\FakeItEasy\Source\FakeItEasy\Core\TypeCatalogueInstanceProvider.cs:line 35
  InnerException: 
```

**Comment by adamralph** [comment]
I see from the source we are try...catching the creation of objects and explicitly suppressing the code analysis warning regarding catch general exception types, so the first chance exceptions are by design. Is this causing any problems?

**Comment by blairconrad** [comment]
Sorry, do you mean "are the first class exceptions causing any problems?"?
Yeah. When debugging in "break on first chance exceptions" mode, users are interrupted with the "there was a first chance exception" dialogue.
I believe this is causing https://github.com/FakeItEasy/FakeItEasy/issues/33#issuecomment-164371645.

**Comment by adamralph** [comment]
The use of exception catching by design is ugly, but sometimes it's the best and/or only way. If a users settings say "break on all exception types" when thrown, they are explicitly saying that when any dependency uses this pattern, they want to break. E.g. when using Nancy, there are *plenty* of first chance exceptions thrown during the handling of a vanilla request. I'm not sure we should commit to maintaining some complex code which avoids these first chance exceptions. Of course, if there is an simple way, why not...

**Comment by blairconrad** [comment]
I see your points about not being able to protect everything.

On the other hand, we're taking types pass the `CanBeInstantiatedAs` filter and trying to instantiate them. I think it's reasonable to make `CanBeInstantiatedAs` do a slightly better job in this case (probably by checking [IsGenericTypeDefinition](https://msdn.microsoft.com/en-us/library/system.type.isgenerictypedefinition(v=vs.110).aspx).
Especially when the `TypeCatalogueInstanceProvider` is being tripped up on a _non-instantiable class that FakeItEasy defines_ (not in the test that we have listed above, but when it chokes on `FakeItEasy.DummyDefinition<T>` like in https://github.com/FakeItEasy/FakeItEasy/issues/33#issuecomment-164371645).
If it were just choking on some non-instantiable type that a user had defined, I'd be less concerned. But still, if we absolutely know that a type can't be instantiated, why let it through the filter if it's even a little annoying?

Besides, #33 is a trouble issue for us anyhow, so if we can reduce any behaviour that will lead us back to it and cause confusion about what's going on, I think it's worth the effort.

**Comment by adamralph** [comment]
:+1:

**Comment by thomaslevesque** [comment]
Couldn't we just document that the extension points must satisfy some requirements, and enforce those requirements in the code? i.e. we could require that the types used as extension point must:

- not be abstract
- not be generic
- not be static
- have a public parameterless constructor

This should take care of most cases

**Comment by blairconrad** [comment]
That was the general idea, yeah. `abstract` is already handled by #494. This issue was to handle the open generics.
I don't really mind forcing a public constructor. Do you think there's a benefit in allowing non-public constructors?
And, yeah, updating the docs to be explicit about the requirements is a good idea.

**Comment by thomaslevesque** [comment]
> I don't really mind forcing a public constructor. Do you think there's a benefit in allowing non-public constructors?

Not really. Forcing a public constructor is fine.

**Comment by adamralph** [comment]
> Couldn't we just document....

As an alternative to fixing this 'bug' (I'm not convinced it's a bug) or in addition?

**Comment by thomaslevesque** [comment]
> As an alternative to fixing this 'bug' (I'm not convinced it's a bug) or in addition?

In addition

**Comment by blairconrad** [comment]
> (I'm not convinced it's a bug) 

Totally a bug. A minor one, sure, but still a bug.
FakeItEasy includes within itself classes that, when FakeItEasy looks for extension points, causes FakeItEasy to throw an exception exception.
Since we absolutely know that there are non-instantiable types that we're going to try to instantiate, we should filter them out.

**Comment by adamralph** [comment]
But they are first chance exceptions, correct? If so, I don't see how that can be a bug.

**Comment by blairconrad** [comment]
I suppose it depends on what we think a "bug" is.

The symptom this issue was intended to correct is unintended behaviour, which I'd argue is undesirable.

Any exception that's thrown will cause the runtime to fire up some exception-handling mechanisms which could have a number of unwanted effects, such as bothering people who break on first chance exceptions, or just slowing things down.

If we want to say that removing the unintended undesirable behaviour that _may_ bother the user is closer to a performance improvement (and thus enhancement) than a bug, I'm happy.

**Comment by adamralph** [comment]
> bothering people who break on first chance exceptions

Then they shouldn't break on first chance exceptions. As I said above, there are many libraries out there which raise first chance exceptions during regular execution.

>slowing things down

Yes, that's a valid reason to change this behaviour.

On that basis, I'm changing this to an enhancement.

**Comment by adamralph** [comment]
Added to 3.1.0 milestone.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 3.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.1.0).
